File: sdks/go/query.go
package modula

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
	"strconv"
)

// QueryResource provides filtered, sorted, paginated content queries by datatype.
// This is the primary way to fetch content for frontend rendering, supporting
// field-level filtering, locale selection, status filtering, and pagination.
// It is accessed via [Client].Query.
type QueryResource struct {
	http *httpClient
}

// QueryParams configures filtering, sorting, and pagination for a content query.
// All fields are optional; zero values are omitted from the request.
type QueryParams struct {
	// Sort specifies the sort field and direction (e.g., "date_created", "-date_modified").
	// Prefix with "-" for descending order.
	Sort string
	// Limit is the maximum number of items to return. Server default applies when 0.
	Limit int
	// Offset is the number of items to skip for pagination. Starts at 0.
	Offset int
	// Locale filters results to content with translations in the given locale code
	// (e.g., "en-US", "fr-FR"). When empty, the default locale is used.
	Locale string
	// Status filters results by content status (e.g., "published", "draft").
	// When empty, all statuses are returned.
	Status string
	// Filters is a map of field-name to value for field-level filtering.
	// Keys correspond to field names defined on the datatype; values are
	// matched exactly against content field values.
	Filters map[string]string
}

// QueryResult is the paginated response envelope for a content query.
type QueryResult struct {
	// Data contains the content items matching the query, up to Limit items.
	Data []QueryItem `json:"data"`
	// Total is the total number of items matching the query across all pages.
	Total int64 `json:"total"`
	// Limit is the page size used for this query.
	Limit int64 `json:"limit"`
	// Offset is the number of items skipped before this page.
	Offset int64 `json:"offset"`
	// Datatype contains metadata about the datatype that was queried.
	Datatype QueryDatatype `json:"datatype"`
}

// QueryItem represents a single content item in a query result, with its
// field values flattened into a string map keyed by field name.
type QueryItem struct {
	// ContentDataID is the unique ULID of this content item.
	ContentDataID string `json:"content_data_id"`
	// DatatypeID is the ULID of the datatype this content belongs to.
	DatatypeID string `json:"datatype_id"`
	// AuthorID is the ULID of the user who created this content.
	AuthorID string `json:"author_id"`
	// Status is the content's lifecycle status ("draft", "published", "scheduled").
	Status string `json:"status"`
	// DateCreated is the ISO 8601 timestamp when the content was created.
	DateCreated string `json:"date_created"`
	// DateModified is the ISO 8601 timestamp when the content was last modified.
	DateModified string `json:"date_modified"`
	// PublishedAt is the ISO 8601 timestamp when the content was published.
	// Empty if the content has never been published.
	PublishedAt string `json:"published_at"`
	// Fields is a map of field name to field value for this content item.
	// All values are strings; parse them according to the field type defined
	// on the datatype schema.
	Fields map[string]string `json:"fields"`
}

// QueryDatatype holds the datatype metadata included in a [QueryResult].
type QueryDatatype struct {
	// Name is the datatype's programmatic name (slug), used in API paths.
	Name string `json:"name"`
	// Label is the datatype's human-readable display name.
	Label string `json:"label"`
}

// Query fetches a filtered, sorted, paginated list of content items for the given
// datatype name (slug). The datatype parameter is required and must match an existing
// datatype's Name field. Pass nil for params to use server defaults (no filters,
// default sort and pagination).
//
// Example:
//
//	result, err := client.Query.Query(ctx, "blog-posts", &modula.QueryParams{
//	    Sort:   "-date_created",
//	    Limit:  10,
//	    Status: "published",
//	})
func (r *QueryResource) Query(ctx context.Context, datatype string, params *QueryParams) (*QueryResult, error) {
	if datatype == "" {
		return nil, fmt.Errorf("modula: datatype name is required")
	}
	p := url.Values{}
	if params != nil {
		if params.Sort != "" {
			p.Set("sort", params.Sort)
		}
		if params.Limit > 0 {
			p.Set("limit", strconv.Itoa(params.Limit))
		}
		if params.Offset > 0 {
			p.Set("offset", strconv.Itoa(params.Offset))
		}
		if params.Locale != "" {
			p.Set("locale", params.Locale)
		}
		if params.Status != "" {
			p.Set("status", params.Status)
		}
		for k, v := range params.Filters {
			p.Set(k, v)
		}
	}
	var raw json.RawMessage
	if err := r.http.get(ctx, "/api/v1/query/"+datatype, p, &raw); err != nil {
		return nil, fmt.Errorf("query %s: %w", datatype, err)
	}
	var result QueryResult
	if err := json.Unmarshal(raw, &result); err != nil {
		return nil, fmt.Errorf("unmarshal query result: %w", err)
	}
	return &result, nil
}

File: sdks/typescript/types/src/entities/query.ts
/**
 * Query parameter types for the ModulaCMS content query API.
 *
 * The query system provides filtered, sorted, paginated access to content
 * of a specific datatype. Queries are scoped to a single datatype (identified
 * by slug in the URL path) and return content items with their field values
 * flattened into a key-value map.
 *
 * @remarks
 * All query parameters are optional. Omitting all parameters returns the
 * first 20 published content items in default sort order.
 *
 * @module entities/query
 */

/**
 * Parameters for a content query request.
 *
 * Passed as query string parameters to `GET /api/v1/query/{datatype_slug}`.
 *
 * @remarks
 * **Filter syntax:** The `filters` record supports both exact-match and operator-based
 * filtering. For exact match, use the field name as the key. For operator-based filtering,
 * use the format `field[op]` where `op` is one of:
 * - `eq` -- equal (same as exact match)
 * - `ne` -- not equal
 * - `gt` -- greater than
 * - `gte` -- greater than or equal
 * - `lt` -- less than
 * - `lte` -- less than or equal
 * - `like` -- SQL LIKE pattern match (use `%` as wildcard)
 * - `in` -- comma-separated list of values
 *
 * **Sort syntax:** Prefix the field name with `-` for descending order. Only one
 * sort field is supported per request.
 *
 * **Pagination:** Use `limit` and `offset` for cursor-free pagination.
 * The response includes `total` for calculating page counts.
 *
 * @example
 * ```ts
 * // Fetch the 10 most recently published blog posts
 * const params: QueryParams = {
 *   sort: '-published_at',
 *   limit: 10,
 *   status: 'published',
 * }
 * ```
 *
 * @example
 * ```ts
 * // Filter by category with pagination
 * const params: QueryParams = {
 *   filters: { category: 'tutorials' },
 *   limit: 20,
 *   offset: 40, // page 3
 * }
 * ```
 *
 * @example
 * ```ts
 * // Operator-based filtering: posts with views > 100
 * const params: QueryParams = {
 *   filters: { 'views[gt]': '100' },
 *   sort: '-views',
 * }
 * ```
 *
 * @example
 * ```ts
 * // Multi-locale query for French content
 * const params: QueryParams = {
 *   locale: 'fr',
 *   status: 'published',
 * }
 * ```
 */
export interface QueryParams {
  /**
   * Sort field name. Prefix with `-` for descending order.
   *
   * Sortable fields include any field defined on the datatype, plus the
   * built-in fields `date_created`, `date_modified`, and `published_at`.
   *
   * @example `'title'` -- ascending by title
   * @example `'-published_at'` -- descending by publish date (newest first)
   */
  sort?: string
  /**
   * Maximum number of items to return.
   *
   * @defaultValue `20`
   * @remarks Clamped to a server-side maximum of `100`. Values above 100 are silently reduced.
   */
  limit?: number
  /**
   * Number of items to skip for offset-based pagination.
   *
   * @defaultValue `0`
   * @remarks Use in combination with {@link QueryResult.total} to calculate page boundaries.
   */
  offset?: number
  /**
   * Locale code for filtering internationalized content (e.g. `'en'`, `'fr'`, `'de'`).
   *
   * When set, only field values matching this locale are included in the response.
   * When omitted, the server returns the default locale.
   */
  locale?: string
  /**
   * Content status filter.
   *
   * Common values: `'published'`, `'draft'`, `'archived'`.
   *
   * @defaultValue `'published'`
   */
  status?: string
  /**
   * Field filters as key-value pairs.
   *
   * Keys are field names, optionally suffixed with an operator in brackets.
   * Values are always strings (the server handles type coercion).
   *
   * @remarks
   * Supported operator suffixes: `[eq]`, `[ne]`, `[gt]`, `[gte]`, `[lt]`, `[lte]`, `[like]`, `[in]`.
   * A bare field name (no operator) is treated as `[eq]`.
   *
   * @example `{ category: 'news' }` -- exact match
   * @example `{ 'price[gte]': '10', 'price[lte]': '50' }` -- range filter
   * @example `{ 'tag[in]': 'go,rust,zig' }` -- match any of the listed values
   * @example `{ 'title[like]': '%tutorial%' }` -- pattern match
   */
  filters?: Record<string, string>
}

/**
 * A single content item in a query result.
 *
 * Represents one row of content data with its field values flattened into
 * a string-keyed map. Field keys correspond to the `name` property of the
 * field definitions on the content's datatype.
 *
 * @remarks
 * All field values are serialized as strings regardless of their underlying
 * {@link FieldType}. Consumers should parse numeric, boolean, JSON, and date
 * field values according to the field type metadata from the schema API.
 */
export interface QueryItem {
  /** ULID of this content data record. */
  content_data_id: string
  /** ULID of the datatype that defines this content's schema. */
  datatype_id: string
  /** ULID of the user who created this content. */
  author_id: string
  /** Publication lifecycle status (e.g. `'published'`, `'draft'`, `'archived'`). */
  status: string
  /** ISO 8601 creation timestamp. */
  date_created: string
  /** ISO 8601 last-modification timestamp. */
  date_modified: string
  /** ISO 8601 timestamp when this content was published, or empty string if unpublished. */
  published_at: string
  /**
   * Field values keyed by field name.
   *
   * Keys are the `name` property of the datatype's field definitions.
   * Values are always strings; parse according to the field's {@link FieldType}.
   *
   * @example `{ title: 'Hello World', body: '<p>Content</p>', views: '42' }`
   */
  fields: Record<string, string>
}

/**
 * Datatype metadata included in a query result.
 *
 * Provides the identity of the datatype that was queried, so consumers
 * can display or route based on the content type without a separate
 * schema lookup.
 */
export interface QueryDatatype {
  /** Machine-readable datatype name (matches the slug used in the query URL). */
  name: string
  /** Human-readable display label for this datatype. */
  label: string
}

/**
 * Response envelope for a content query.
 *
 * Contains the matched content items, pagination metadata, and the
 * identity of the queried datatype.
 *
 * @remarks
 * Use {@link total}, {@link limit}, and {@link offset} to implement
 * pagination controls. The total number of pages is `Math.ceil(total / limit)`.
 *
 * @example
 * ```ts
 * const result: QueryResult = await client.query('blog-posts', {
 *   sort: '-published_at',
 *   limit: 10,
 * })
 *
 * console.log(`Showing ${result.data.length} of ${result.total} items`)
 * const totalPages = Math.ceil(result.total / result.limit)
 * const currentPage = Math.floor(result.offset / result.limit) + 1
 * ```
 */
export interface QueryResult {
  /** Array of content items matching the query. May be empty if no results match. */
  data: QueryItem[]
  /** Total number of items matching the query (ignoring limit/offset). */
  total: number
  /** The limit that was applied to this query (echoed from the request, or the default). */
  limit: number
  /** The offset that was applied to this query (echoed from the request, or 0). */
  offset: number
  /** Metadata about the datatype that was queried. */
  datatype: QueryDatatype
}
